### ISSUE TITLE
how to change to generator to use a custom data array of time series data

### ISSUE THREAD

**Issue by wilbertmatthew** [fabridamicelli/echoes#1]
Hi,

I am new to echoes and would like to use a data array of time series data that I collect as the generator data feed. How would I change from the MackeyGlass data to my own heart rate data time series? I tried just replacing the MackeyGlass data but the prediction results are way off. 

Thanks

Wilbert Jackson

**Comment by fabridamicelli** [comment]
Hi,
It'd be great if you could post a code snippet showing an concrete example so that I can better understand what exactly you're trying to achieve and see if I can help.

Best
F

**Comment by wilbertmatthew** [comment]
Fabri thanks for your response and excellent work. What I am trying to do is use the generator as a framework for multi step prediction of different time series data sets. The code included is your generator example modified to run different data sets. I ran three different generated Mackey Glass data sets and each produced good predictions. I ran a Sine Wave equation generated data set and Atrial Fibrillation file generated data set which both produce bad predictions. Its seems the generator works well with any Mackey Glass type oscillation but not with any other type. What am I doing wrong?

Thanks

import numpy as np
from matplotlib import pyplot as plt
from sklearn.metrics import r2_score
from sklearn.model_selection import train_test_split

from echoes import ESNGenerator
from echoes.datasets import load_mackeyglasst17
from echoes.plotting import plot_predicted_ts, set_mystyle
from sklearn.preprocessing import MinMaxScaler

set_mystyle()  # optional, set aesthetics

# Load MackeyGlass data 3 diiferent ways. All predictions are good in the range of 0.9999994836472421
################################## 1
mackey_ts = load_mackeyglasst17()
mackey_ts = np.array(mackey_ts)
mackey_ts = mackey_ts.reshape(-1, 1)
##################################

################################## 2
N  = 10000
b   = 0.1
c   = 0.2
tau = 18

mackey_ts = [0.9697, 0.9699, 0.9794, 1.0003, 1.0319, 1.0703, 1.1076, 1.1352, 1.1485,
     1.1482, 1.1383, 1.1234, 1.1072, 1.0928, 1.0820, 1.0756, 1.0739, 1.0759]

for n in range(17,N+99):
    mackey_ts.append(mackey_ts[n] - b*mackey_ts[n] + c*mackey_ts[n-tau]/(1+mackey_ts[n-tau]**10))

mackey_ts = mackey_ts[50:]
mackey_ts = np.array(mackey_ts)
mackey_ts = mackey_ts.reshape(-1, 1)
##################################

################################## 3
mackey_ts = np.loadtxt('mackeyglass1.txt')
mackey_ts = np.array(mackey_ts)
mackey_ts = mackey_ts.reshape(-1, 1)
##################################

#Create a Sine Wave. Prediction is bad.
################################## 1
mackey_ts = np.sin(np.linspace(0, 18*np.pi, 5000)).reshape(-1, 1)
mackey_ts = np.array(mackey_ts)
mackey_ts = mackey_ts.reshape(-1, 1)
##################################

#Load Atrial Fibrillation. Prediction is bad.
##################################
mackey_ts = np.loadtxt('afib.txt')
mackey_ts = np.array(mackey_ts)
mackey_ts = mackey_ts.reshape(-1, 1)
# Tried scaling
#mackey_ts = 2 * (mackey_ts - mackey_ts.min()) / (mackey_ts.max() - mackey_ts.min()) - 1
##################################

n_train_steps, n_test_steps = 2000, 150
n_total_steps = n_train_steps + n_test_steps

y_train, y_test = train_test_split(
    mackey_ts, 
    train_size=n_train_steps, 
    test_size=n_test_steps, 
    shuffle=False
)

esn = ESNGenerator(
    n_steps=n_test_steps,
    n_reservoir=200,
    spectral_radius=1.25,   
    leak_rate=.4,
    random_state=42,
)
# Fit the model. Inputs is None because we only have the target time series
esn.fit(X=None, y=y_train)

y_pred = esn.predict()
print("test r2 score", r2_score(y_test, y_pred))

# Plot training and test
plt.figure(figsize=(22, 5))
plt.plot(mackey_ts[: n_total_steps], 'steelblue', linewidth=5, label="target system")
plt.plot(esn.training_prediction_, color="y", linewidth=1, label="training fit")
plt.plot(range(n_train_steps, n_total_steps), y_pred,'orange', label="ESNGenerator")
plt.ylabel("oscillator value")
plt.xlabel('time')
lo, hi = plt.ylim()
plt.vlines(n_train_steps, lo-.05, hi+.05, linestyles='--')
plt.legend(fontsize='small')

# Plot test alone
plt.figure(figsize=(22, 5))
plt.subplot(1, 4, (1, 3))
plt.title("zoom into test")
plt.plot(y_test,
         color="steelblue",
         label="target system", 
         linewidth=5.5)
plt.xlabel('time')

plt.plot(y_pred, 
         linestyle='--',
         color="orange", 
         linewidth=2,
         label="generative ESN",)
plt.ylabel("oscillator")
plt.xlabel('time')
plt.legend(fontsize='small')

plt.subplot(1, 4, 4)
plt.title(r"$W^{out}$ weights distribution")
plt.xlabel('weight')
plt.ylabel('frequency')
plt.hist(esn.W_out_.flat)
plt.tight_layout();

plt.show();

[afib.txt](https://github.com/user-attachments/files/15521533/afib.txt)

**Comment by wilbertmatthew** [comment]
Fabrizio,

Sorry I forgot to include the mackeyglass1.txt file data set.
[mackeyglass1.txt](https://github.com/user-attachments/files/15521713/mackeyglass1.txt)

**Comment by fabridamicelli** [comment]
@wilbertmatthew Thank you for the code and the data.
I looked into it  and managed to reproduce your results.
At first glance it looks like a bug in the library to me. But I need a bit more time to investigate/fix it, so I'll get back to you soon.

**Comment by wilbertmatthew** [comment]
Thanks so much for efforts and response.

**Comment by wilbertmatthew** [comment]
Hi,
I have been looking into the problem with bad predictions on data other than the mackeyglass data as a way to familiarize myself with the code. I have been using the afib.txt file I sent you. I may be off base but I what noticed is that datasets that have whole numbers or whole number decimal part do not produce the correct states through the harvest_states function. I have included some plots.

1. mackeyglassstates.png which shows the states of the mackeyglass dataset.
2. mackeyglassprediction.png
3. mackeyglassweights.png
 
4. afibbadstates.png shows the states when afib.txt has mix of whole numbers and floating point numbers.
5. afibbadprediction.png the prediction for the above states.
6. afibbadweights.png  the weights for above states.

7. afibstatesdecimalnumbers.png shows the afib.txt states when I converted the afib data to all decimal numbers.
8. afibpredictionwithstatesasdecimalsonly.png shows the prediction for afib.txt as all decimal numbers.
9. afibweightsasdecimalnumbers.png shows the weights for afib.txt as all decimal numbers.

![mackeyglassweights](https://github.com/fabridamicelli/echoes/assets/2045731/5d065e5b-c37b-4893-b4dd-8073a24bb39b)
![mackeyglassstates](https://github.com/fabridamicelli/echoes/assets/2045731/9bbcbe08-d143-4af2-89b9-e98c79e3f187)
![mackeyglassprediction](https://github.com/fabridamicelli/echoes/assets/2045731/7c03eabd-22ce-49df-be5e-c1ebd3414bad)
![afibweightsasdecimalnumbers](https://github.com/fabridamicelli/echoes/assets/2045731/a51de24a-1e11-4068-97c2-477c505fdfde)
![afibstatesasdecimalnumbers](https://github.com/fabridamicelli/echoes/assets/2045731/56edf402-f28f-4f3c-8400-18ff2b521177)
![afibpredictionwithstatesasdecimalsonly](https://github.com/fabridamicelli/echoes/assets/2045731/bc036a22-4eee-4415-a23b-af7cba62a4cc)
![afibbadweights](https://github.com/fabridamicelli/echoes/assets/2045731/15ed4b2d-e62d-4cc6-983a-4504b0e9bf2c)
![afibbadstates](https://github.com/fabridamicelli/echoes/assets/2045731/c5ca1bb9-15f0-4df7-b8ca-0b86ab9da0cf)
![afibbadprediction](https://github.com/fabridamicelli/echoes/assets/2045731/ae5abc79-3ca9-4935-898c-25f3e9dc27cb)

**Comment by fabridamicelli** [comment]
Hi @wilbertmatthew 
thank you for taking the lead and moving on with the experiments to find the bug –  way to go!
I do have this issue on my TODO list,I will take a closer look at it as soon as I can!

**Comment by wilbertmatthew** [comment]
Thank you.

**Comment by fabridamicelli** [comment]
Hi @wilbertmatthew 

sorry for the long delay.
I've been looking at the problem a bit closer and I have a question:
What do you exactly mean by converting the numbers to decimal? It would be great if you could send me the code with that you used for the transformation for me to reproduce the results.
Also, I tried playing around with the hyperparameters and found some interesting differences, in particular adjusting the number of neurons and spectral radius (see code below). But the whole system is still very unstable with respect to those hyperparameters.

[Here's a notebook](https://github.com/fabridamicelli/echoes/blob/master/examples/notebooks/plot_generator_atrial.ipynb) with some results I got using your data which look a bit more normal.
Please take a look at that and let me know what you think

Best,
F

**Comment by wilbertmatthew** [comment]
Thank you very much for taking the time to look at the issue. I will play
around with your version and see if the number conversion was even an
issue. Could have been something I was doing wrong in my attempt to get
some reasonable predictions using the afib time series.

Wilbert

**Comment by fabridamicelli** [comment]
@wilbertmatthew any news on this? Otherwise I will close the issue as I am trying to tidy up things for a new release

**Comment by wilbertmatthew** [comment]
I have not been able to improve the results any better than what you have
provided. So I guess it's a closed issue. Thanks.

**Comment by fabridamicelli** [comment]
@wilbertmatthew Thank you very much for your valuable feedback.

I am closing this issue, but please feel free to open a new one if you find any other issues with library.
